### DISCUSSION TITLE
OSError: [Errno 116] ETIMEDOUT

### DISCUSSION THREAD

**Discussion by Allan196** [micropython/micropython#12888] [ESP32]
Platform: ESP32-WROVER (firmware esp32spiram-20220618-v1.19.1.bin) 
After connecting to a WiFi network, this code snippet works as expected but sometimes results in an exception "OSError: [Errno 116] ETIMEDOUT"
Why doesn't this work 100% of the time?

```py
    try: 
      conn, addr = s.accept()
    except:
      pass # no Request from client
    else:
      # Received a connection request
      try: 
        request_str = conn.recv(2048)
      except:
        # OSError: [Errno 116] ETIMEDOUT
```

**Comment by peterhinch** [comment]
WiFi connections of "Bare metal" hardware are not 100% reliable. They are vulnerable to transient disruption from radio interference. This inherent unreliability is masked by the OS of a PC which gives the illusion of a rock-solid connection. In application design it's wise to include error trapping at every stage.

**Reply by Allan196** [reply]
Thanks for your response.
When I change my code to no timeout ie wait forever and then execute "request_str = conn.recv(2048)", for those occasions when there is say a "transient disruption from radio interference" a return from the function call never happens. Instead the browser forever displays "Waiting for 192.168.1.120". It would be nice to know what is actually happening in the library function "conn.recv".

**Reply by glenn20** [reply]
https://github.com/micropython/micropython/blob/master/ports/esp32/modsocket.c#L499
and
https://github.com/espressif/esp-lwip/blob/6bf7044c0368d587f70f7083e39e0a619b4d5788/src/api/sockets.c#L1205

cover the details. This reaches down into the espressif and lwip network stack. Does your client attempt to reconnect if the first attempt fails?

Without further details, I woud recommend you retain the timeout for robust operation and your client should attempt to reconnect on failure. Just treat failed connections as part of normal operations.

**Reply by Allan196** [reply]
The client never has to re-connect to the WiFi network (that is rock solid once the initial connection is made), but instead in the browser window, again press the "Submit" button when timeouts do occur.
Any tips for bufsiz in conn.recv(bufsiz) relative to how many bytes are expected to be received?

**Reply by glenn20** [reply]
Yes, I was referring to the tcp connections rather than the wifi connections. Re. the bufsiz, you could try setting it much smaller to see if any data is being received to diagnose the problem. You could even try reading one char at a time to try diagnosing the problem.
